RxSwift/Observables/Implementations/SwitchIfEmpty.swift

```diff
@@ -0,0 +1,127 @@
+    }
+}
+
+final class SwitchIfEmptySink<SourceType, S: ObservableConvertibleType, O: ObserverType>: Sink<O>
+    , ObserverType
+    , LockOwnerType
+    , SynchronizedOnType where S.E == O.E, SourceType == S.E {
+    typealias E = SourceType
+    typealias Selector = (Void) throws -> S
+    
+    let _lock = NSRecursiveLock()
+    private let _selector: Selector
+    fileprivate var isEmpty = true
+    fileprivate let _subscriptions: SingleAssignmentDisposable = SingleAssignmentDisposable()
+    fileprivate let _innerSubscription: SerialDisposable = SerialDisposable()
+    
+    init(selector: @escaping Selector, observer: O, cancel: Cancelable) {
+        _selector = selector
+        super.init(observer: observer, cancel: cancel)
+    }
+    
+    func run(_ source: Observable<SourceType>) -> Disposable {
+        let subscription = source.subscribe(self)
+        _subscriptions.setDisposable(subscription)
+        return Disposables.create(_subscriptions, _innerSubscription)
+    }
+    
+    func on(_ event: Event<E>) {
+        synchronizedOn(event)
+    }
+    
+    func _synchronized_on(_ event: Event<E>) {
+        switch event {
+        case .next:
+            isEmpty = false
+            forwardOn(event)
+        case .error:
+            forwardOn(event)
+            dispose()
+        case .completed:
+            if isEmpty {
+                do {
+                    let observable = try _selector().asObservable()
+                    let d = SingleAssignmentDisposable()
+                    _innerSubscription.disposable = d
+                    
+                    let observer = SwitchIfEmptySinkIter(parent: self, _self: d)
+                    let disposable = observable.subscribe(observer)
+                    d.setDisposable(disposable)
+                } catch let error {
+                    forwardOn(.error(error))
+                    dispose()
+                }
+            } else {
+                forwardOn(.completed)
+                dispose()
+            }
+        }
+    }
+}
+
+final class SwitchIfEmptySinkIter<SourceType, S: ObservableConvertibleType, O: ObserverType>
+    : ObserverType
+    , LockOwnerType
+    , SynchronizedOnType where S.E == O.E, SourceType == S.E {
+    typealias E = SourceType
+    typealias Parent = SwitchIfEmptySink<SourceType, S, O>
+    
+    let _lock = NSRecursiveLock()
+    fileprivate let _parent: Parent
+    fileprivate let _self: Disposable
+    
+    init(parent: Parent, _self: Disposable) {
+        _parent = parent
+        self._self = _self
+    }
+    
+    func on(_ event: Event<E>) {
+        synchronizedOn(event)
+    }
+    
+    func _synchronized_on(_ event: Event<E>) {
+        switch event {
+        case .next:
+            _parent.forwardOn(event)
+        case .error:
+            _parent.forwardOn(event)
+            _self.dispose()
+            _parent.dispose()
+        case .completed:
+            _parent.isEmpty = false
```

Comment:
Hm. But would it try to switch again, 
e.g. In case where `source` and `switch` are both empty? Or I missing something?